prql-compiler/src/semantic/type_resolver.rs

```diff
@@ -91,12 +91,24 @@ where
         Some(Ordering::Equal | Ordering::Greater)
     );
     if !expected_is_above {
-        return Err(Error::new(Reason::Expected {
+        let e = Err(Error::new(Reason::Expected {
             who: who(),
             expected: format!("type `{}`", expected),
             found: format!("type `{}`", found_ty),
         })
         .with_span(found.span));
+        if matches!(found_ty, Ty::Function(_)) && !matches!(expected, Ty::Function(_)) {
+            return e.with_help(match &found.kind {
+                ExprKind::Closure(closure) => match &closure.name {
```

Comment:
If you prefer (but no need to change, let's merge regardless), we can encapsulate the initial conditions in a single statement, I think, and then only have two cases:

```suggestion
                ExprKind::Closure(Some(name)) => {
```

...though possibly the references don't quite work and we need the three conditions.

Reply:
I don't think this is possible because `ExprKind::Closure` contains a `Box` which prevents further pattern matching.

Reply:
Ah thanks @aljazerzen. 

Sorry for the noise @JettChenT ! Ignore my comment. 

(We can revisit on https://doc.rust-lang.org/beta/unstable-book/language-features/box-patterns.html 😄  )

Reply:
I originally tried to use the if let guard (https://github.com/rust-lang/rust/issues/51114), but it seems as though the feature is still experimental.